Userland/Utilities/sed.cpp

```diff
@@ -14,138 +20,866 @@
+    {
+        return parsing_error(lexer, "{}", message);
+    }
+
+    template<typename... Parameters>
+    static SedError parsing_error(GenericLexer const& lexer, CheckedFormatString<Parameters...>&& fmtstr, Parameters const&... parameters)
+    {
+        StringBuilder builder;
+        builder.appendff("Parsing error at position {}: ", lexer.tell());
+        builder.appendff(move(fmtstr), parameters...);
+        return maybe_with_string(String::from_utf8(builder.string_view()));
+    }
+
+    static SedError from_error(Error const& error)
+    {
+        return formatted("Internal sed error: {}", error.string_literal());
+    }
+
+private:
+    String m_message;
+
+    static SedError maybe_with_string(ErrorOr<String> maybe_string)
+    {
+        if (maybe_string.is_error())
+            return SedError {};
+        return SedError { maybe_string.release_value() };
+    }
+};
+
+template<typename T>
+using SedErrorOr = ErrorOr<T, SedError>;
+
+// function, name, maximum addresses
+#define ENUMERATE_FUNCTIONS(F) \
+    F('a', A, 1)               \
+    F('b', V, 2)               \
+    F('c', C, 2)               \
+    F('d', d, 2)               \
+    F('D', D, 2)               \
+    F('g', g, 2)               \
+    F('G', G, 2)               \
+    F('h', h, 2)               \
+    F('H', H, 2)               \
+    F('i', I, 1)               \
+    F('l', l, 2)               \
+    F('n', n, 2)               \
+    F('N', N, 2)               \
+    F('p', p, 2)               \
+    F('P', P, 2)               \
+    F('q', q, 1)               \
+    F('r', R, 1)               \
+    F('s', S, 2)               \
+    F('t', T, 2)               \
+    F('w', W, 2)               \
+    F('x', x, 2)               \
+    F('y', Y, 2)               \
+    F(':', Colon, 0)           \
+    F('=', equals, 1)
+
+enum class AddressType {
+    Unset,
+    Line,
+    LastLine,
+    ContextAddress,
+};
+
+class Address {
+public:
+    Address() = default;
+
+    explicit Address(size_t line)
+        : m_line_number(line)
+        , m_address_type(AddressType::Line)
+    {
+    }
+
+    explicit Address(AddressType address_type)
+        : m_address_type(address_type)
+    {
+        VERIFY(address_type == AddressType::LastLine || address_type == AddressType::ContextAddress);
+    }
+
+    size_t line_number() const
+    {
+        VERIFY(m_address_type == AddressType::Line);
+        return m_line_number;
+    }
+
+    AddressType address_type() const { return m_address_type; }
+
+    bool matches([[maybe_unused]] StringView pattern_space, size_t line_number, bool is_last_line) const
+    {
+        switch (m_address_type) {
+        case AddressType::Line:
+            return line_number == m_line_number;
+        case AddressType::LastLine:
+            return is_last_line;
+        default:
+            Core::File::standard_error()->write(DeprecatedString::formatted("Addressing type not implemented: {}", int(m_address_type)));
```

Comment:
```suggestion
            warnln("sed: Addressing type not implemented: {}", to_underlying(m_address_type));
```

Reply:
Thanks @alimpfard, I've changed all stderr calls to use `warnln`, and most stdout calls to use `outln` (I'm still passing the stdout `File` object around in some places explicitly as sometimes outputs goes into a file instead).